thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,163 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef TRACESLAVE_HPP
+#define TRACESLAVE_HPP
+
+#include "slave.hpp"
+#include "jqueue.tpp"
+#include "jthread.hpp"
+#include "limits.h"
+
+
+class CTracingThorDataLink :  implements CInterfaceOf<IThorDataLink>
+{
+private:
+    class TraceQueue
+    {
+    private:
+        void dump(MemoryBuffer &mb, IHThorArg *helper) const
+        {
+            // NOTE - we don't output every item in the list to be 
+            // consistent with the other dump() implementation
+            // Note - whilst dumping, does not clear ring buffer
+            // so that it is available later
```

Comment:
The rowLogCache, is only ever in use by 1 thread, so doesn't need to be +1 in size and skip the rowBuffer[rowHeadp]'th element.
I think this should be sized to N (e.g. default 10) and return all.

It will make it clearer that different and not in conflict like the other buffers,